src/main/java/org/folio/service/organization/OrganizationStorageService.java

```diff
@@ -77,112 +79,122 @@ private boolean isSameAccountNumbers(Organization organization) {
-  public CompletableFuture<Void> updateOrganizationById(String id, Organization updatedOrganization, String lang, Context context,
+  public Future<Void> updateOrganizationById(String id, Organization updatedOrganization, String lang, Context context,
       Map<String, String> headers) {
     logger.debug("updateOrganization:: Trying to update organization with id: {}", id);
-    CompletableFuture<Void> future = new CompletableFuture<>();
+
+    Promise<Void> promise = Promise.promise();
+    Promise<Organization> organizationPromise = Promise.promise();
+    RequestContext requestContext = new RequestContext(context, headers);
+
     if (isEmpty(updatedOrganization.getId())) {
       updatedOrganization.setId(id);
     } else if (!id.equals(updatedOrganization.getId())) {
       logger.warn("updateOrganization:: Mismatch between id '{}' in path and request body '{}'", id, updatedOrganization.getId());
-      future.completeExceptionally(new HttpException(HttpStatus.HTTP_UNPROCESSABLE_ENTITY.toInt(),
+      return Future.failedFuture(new HttpException(HttpStatus.HTTP_UNPROCESSABLE_ENTITY.toInt(),
         MISMATCH_BETWEEN_ID_IN_PATH_AND_BODY.toError()));
-      return future;
     }
     if (isSameAccountNumbers(updatedOrganization)) {
       logger.warn("updateOrganization:: Account number of organization '{}' is not unique", id);
-      future.completeExceptionally(new HttpException(HttpStatus.HTTP_UNPROCESSABLE_ENTITY.toInt(),
+      return Future.failedFuture(new HttpException(HttpStatus.HTTP_UNPROCESSABLE_ENTITY.toInt(),
         ACCOUNT_NUMBER_MUST_BE_UNIQUE.toError()));
-      return future;
     }
-    HttpClientInterface client = getHttpClient(headers);
-    handleGetRequest(resourceByIdPath(ORGANIZATIONS, id), client, headers, logger)
-      .thenApply(existingOrganizationJson -> existingOrganizationJson.mapTo(Organization.class))
-      .thenCompose(existingOrganization -> protectionService.validateAcqUnitsOnUpdate(updatedOrganization, existingOrganization, lang, context, headers))
-      .thenAccept(ok -> handlePutRequest(resourceByIdPath(ORGANIZATIONS, updatedOrganization.getId()), JsonObject.mapFrom(updatedOrganization), client, headers,
-        logger).handle((org, t) -> {
-        client.closeClient();
+    handleGetRequest(resourceByIdPath(ORGANIZATIONS, id), requestContext, logger)
+      .compose(existingOrganizationJson -> {
+        Organization organization = existingOrganizationJson.mapTo(Organization.class);
+        organizationPromise.complete(organization);
+        return organizationPromise.future();
+      })
+      .compose(existingOrganization -> protectionService.validateAcqUnitsOnUpdate(updatedOrganization, existingOrganization, lang, context, headers)
+      .compose(ok -> handlePutRequest(resourceByIdPath(ORGANIZATIONS, updatedOrganization.getId()), JsonObject.mapFrom(updatedOrganization), logger, requestContext))
+        .compose(ok -> {
+          promise.complete();
+        return promise.future();
```

Comment:
This piece of code will work, but looks confusing. You could get rid of promise as I mentioned above to avoid that

Reply:
fixed